### PR TITLE
Document design of the shm transport

### PR DESCRIPTION
**Pull request by pietern**

### CODE REVIEW

File: docs/shm.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+is automatically cleaned up when the last process that has its file
+descriptor terminates.
+
+Per above, creating a shared memory ring buffer yields 2 file
+descriptors, one for the header segment and one for the data segment.
+These file descriptors are shared over a Unix domain socket.
+
+### The reactor
+
+This is a tensorpipe specific component. It uses a shared memory ring
```
Comment (lerks) [inline docs/shm.md]: s/tensorpipe/TensorPipe/

File: docs/shm.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+Each process receives file descriptors from their peer and initializes
+the corresponding ring buffers. The peer's inbox is referred to as the
+_outbox_. The token to trigger remote readability is referred to as
+the _outbox trigger_.
+
+The connection is now established! Writes are performed by writing
+directly into the outbox and triggering the outbox trigger. The
+trigger wakes up the peer's reactor and executes a function that
+notifies the connection of readability. Subsequently, the connection
+checks if there was a pending read operations, and processes it if so.
```
Comment (lerks) [inline docs/shm.md]: s/operations/operation/

**Comment by pietern** [comment]
Thanks for reviewing. Addressed both comments.

File: docs/shm.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+connection, and a context. Listeners accept connections. Contexts
+create listeners and can connect to remote listeners.
+
+## Concepts
+
+
+### Ring buffers
+
+Shared memory ring buffers are a core building block for the shm
+transport. They are implemented with split control and and data
```
Comment (beauby) [inline docs/shm.md]: split control and ~~and~~ data

File: docs/shm.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+1. (+) It is not necessary to purge stale Unix domain socket files.
+2. (-) These sockets don't have permissions, so any process that has
+   its name can connect.
+
+Read more about abstract domain sockets [here][1] and [here][2].
+
+[1]: http://man7.org/linux/man-pages/man7/unix.7.html
+[2]: https://utcc.utoronto.ca/~cks/space/blog/linux/SocketAbstractNamespace
+
+Once processes have established a Unix domain socket, it used to:
```
Comment (beauby) [inline docs/shm.md]: it **is** used to:

**Pull request closed without merging**